File: bungee/src/main/java/net/impactdev/gts/bungee/messaging/types/RedisBungeeMessenger.java
package net.impactdev.gts.bungee.messaging.types;

import com.imaginarycode.minecraft.redisbungee.RedisBungee;
import com.imaginarycode.minecraft.redisbungee.RedisBungeeAPI;
import com.imaginarycode.minecraft.redisbungee.events.PubSubMessageEvent;
import net.impactdev.gts.api.messaging.IncomingMessageConsumer;
import net.impactdev.gts.api.messaging.Messenger;
import net.impactdev.gts.api.messaging.message.OutgoingMessage;
import net.impactdev.gts.bungee.GTSBungeePlugin;
import net.md_5.bungee.api.plugin.Listener;
import net.md_5.bungee.event.EventHandler;
import org.checkerframework.checker.nullness.qual.NonNull;

public class RedisBungeeMessenger implements Messenger, Listener {

	private static final String CHANNEL = "gts:update";

	private final GTSBungeePlugin plugin;
	private final IncomingMessageConsumer consumer;
	private RedisBungeeAPI redisBungee;

	public RedisBungeeMessenger(GTSBungeePlugin plugin, IncomingMessageConsumer consumer) {
		this.plugin = plugin;
		this.consumer = consumer;
	}

	public void init() {
		this.redisBungee = RedisBungee.getApi();
		this.redisBungee.registerPubSubChannels(CHANNEL);

		this.plugin.bootstrap().proxy().getProxy().getPluginManager().registerListener(this.plugin.bootstrap().proxy(), this);
	}

	@Override
	public void close() {
		this.redisBungee.unregisterPubSubChannels(CHANNEL);
		this.redisBungee = null;

		this.plugin.bootstrap().proxy().getProxy().getPluginManager().unregisterListener(this);
	}

	@Override
	public IncomingMessageConsumer getMessageConsumer() {
		return this.consumer;
	}

	@Override
	public void sendOutgoingMessage(@NonNull OutgoingMessage outgoingMessage) {
		this.redisBungee.sendChannelMessage(CHANNEL, outgoingMessage.asEncodedString());
	}

	@EventHandler
	public void onMessage(PubSubMessageEvent event) {
		if(!event.getChannel().equals(CHANNEL)) {
			return;
		}

		this.consumer.consumeIncomingMessageAsString(event.getMessage());
	}
}

File: bungee/src/main/kotlin/net/impactdev/gts/bungee/messaging/types/RedisBungeeMessenger.kt
package net.impactdev.gts.bungee.messaging.types

import com.imaginarycode.minecraft.redisbungee.RedisBungee
import com.imaginarycode.minecraft.redisbungee.RedisBungeeAPI
import com.imaginarycode.minecraft.redisbungee.events.PubSubMessageEvent
import net.impactdev.gts.api.messaging.IncomingMessageConsumer
import net.impactdev.gts.api.messaging.Messenger
import net.impactdev.gts.api.messaging.message.OutgoingMessage
import net.impactdev.gts.bungee.GTSBungeePlugin
import net.md_5.bungee.api.plugin.Listener
import net.md_5.bungee.event.EventHandler

class RedisBungeeMessenger(
    private val plugin: GTSBungeePlugin?,
    override val messageConsumer: IncomingMessageConsumer
) : Messenger, Listener {
    private var redisBungee: RedisBungeeAPI? = null
    fun init() {
        redisBungee = RedisBungee.getApi()
        redisBungee.registerPubSubChannels(CHANNEL)
        plugin!!.bootstrap.proxy.pluginManager.registerListener(plugin.bootstrap, this)
    }

    override fun close() {
        redisBungee!!.unregisterPubSubChannels(CHANNEL)
        redisBungee = null
        plugin!!.bootstrap.proxy.pluginManager.unregisterListener(this)
    }

    override fun sendOutgoingMessage(outgoingMessage: OutgoingMessage) {
        redisBungee!!.sendChannelMessage(CHANNEL, outgoingMessage.asEncodedString())
    }

    @EventHandler
    fun onMessage(event: PubSubMessageEvent) {
        if (event.channel != CHANNEL) {
            return
        }
        messageConsumer.consumeIncomingMessageAsString(event.message)
    }

    companion object {
        private const val CHANNEL = "gts:update"
    }
}